### ISSUE TITLE
Superset ignore case-sensitivity of time column names with Postgresql

### ISSUE THREAD

**Issue by victornoel** [apache/incubator-superset#5886]
Make sure these boxes are checked before submitting your issue - thank you!

- [x] I have checked the superset logs for python stacktraces and included it here as text if any
- [x] I have reproduced the issue with at least the latest released version of superset
- [x] I have checked the issue tracker for the same issue and I haven't found one similar
 - A closed issue seems to be related: #2588 but was closed because inactivity.

### Superset version

Current master: 85692612d6f1438d8b19f51aefae92a240bc3cf3

### Expected results

A table with column name where case is important (for example `invoiceDate`) should be usable without any problem.

### Actual results

When creating a chart with such a table, I get the following error:
```
column "invoicedate" does not exist LINE 1: SELECT DATE_TRUNC('day', invoiceDate) AT TIME ZONE 'UTC' AS ... ^ HINT: Perhaps you meant to reference the column "InvoiceLine.invoiceDate".
```

It seems to happens because of the `Time Column` case sensitivity in combination with a `Time Grain` (here `day`).

### Steps to reproduce

Import a table with case-sensitive names.

**Comment by villebro** [comment]
@victornoel can you post the full SQL query generated by Superset?

**Comment by victornoel** [comment]
@villebro So I go into charts, I click on the + button, I choose my table where the date column is case sensitive, I choose "time series - line chart" and then I see the error above on the right pane.

I click on show query, and it tells me: "Error..."

And in the logs I have:
```
superset_1  | 2018-09-17 07:26:02,482:INFO:root:Database.get_sqla_engine(). Masked URL: postgresql+psycopg2://user:XXXXXXXXXX@db/db
superset_1  | 2018-09-17 07:26:03,600:ERROR:root:column "invoicedate" does not exist
superset_1  | LINE 1: SELECT DATE_TRUNC('day', invoiceDate) AT TIME ZONE 'UTC' AS ...
superset_1  |                                  ^
superset_1  | HINT:  Perhaps you meant to reference the column "InvoiceLine.invoiceDate".
superset_1  | Traceback (most recent call last):
superset_1  |   File "/home/work/incubator-superset/superset/connectors/sqla/models.py", line 781, in query
superset_1  |     df = self.database.get_df(sql, self.schema)
superset_1  |   File "/home/work/incubator-superset/superset/models/core.py", line 795, in get_df
superset_1  |     self.db_engine_spec.execute(cursor, sqls[-1])
superset_1  |   File "/home/work/incubator-superset/superset/db_engine_specs.py", line 376, in execute
superset_1  |     cursor.execute(query)
superset_1  | psycopg2.ProgrammingError: column "invoicedate" does not exist
superset_1  | LINE 1: SELECT DATE_TRUNC('day', invoiceDate) AT TIME ZONE 'UTC' AS ...
superset_1  |                                  ^
superset_1  | HINT:  Perhaps you meant to reference the column "InvoiceLine.invoiceDate".
superset_1  | 
superset_1  | 2018-09-17 07:26:03,661:INFO:werkzeug:172.22.0.1 - - [17/Sep/2018 07:26:03] "POST /superset/explore_json/?query=true HTTP/1.1" 500 -
superset_1  | Traceback (most recent call last):
superset_1  |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1997, in __call__
superset_1  |     return self.wsgi_app(environ, start_response)
superset_1  |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1985, in wsgi_app
superset_1  |     response = self.handle_exception(e)
superset_1  |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1540, in handle_exception
superset_1  |     reraise(exc_type, exc_value, tb)
superset_1  |   File "/usr/local/lib/python3.6/site-packages/flask/_compat.py", line 33, in reraise
superset_1  |     raise value
superset_1  |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1982, in wsgi_app
superset_1  |     response = self.full_dispatch_request()
superset_1  |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1614, in full_dispatch_request
superset_1  |     rv = self.handle_user_exception(e)
superset_1  |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1517, in handle_user_exception
superset_1  |     reraise(exc_type, exc_value, tb)
superset_1  |   File "/usr/local/lib/python3.6/site-packages/flask/_compat.py", line 33, in reraise
superset_1  |     raise value
superset_1  |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1612, in full_dispatch_request
superset_1  |     rv = self.dispatch_request()
superset_1  |   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1598, in dispatch_request
superset_1  |     return self.view_functions[rule.endpoint](**req.view_args)
superset_1  |   File "/home/work/incubator-superset/superset/models/core.py", line 1010, in wrapper
superset_1  |     value = f(*args, **kwargs)
superset_1  |   File "/usr/local/lib/python3.6/site-packages/flask_appbuilder/security/decorators.py", line 52, in wraps
superset_1  |     return f(self, *args, **kwargs)
superset_1  |   File "/home/work/incubator-superset/superset/views/core.py", line 1174, in explore_json
superset_1  |     force=force)
superset_1  |   File "/home/work/incubator-superset/superset/views/core.py", line 1089, in generate_json
superset_1  |     return self.get_query_string_response(viz_obj)
superset_1  |   File "/home/work/incubator-superset/superset/views/core.py", line 1055, in get_query_string_response
superset_1  |     'data': viz_obj.get_df().to_dict('records'),  # TODO, split into endpoint
superset_1  | AttributeError: 'NoneType' object has no attribute 'to_dict'
```

**Comment by victornoel** [comment]
For the record, I don't have this problem with case-sensitive fields in metrics for example. It seems to only concern the field used for date in relation to the "Time Grain".

**Comment by villebro** [comment]
@victornoel In the upper-right hand corner of the Chart UI, can you click on "Query" - "View Query" and post the contents?

**Comment by victornoel** [comment]
@villebro that's exactly what I did, and the content was: `Error...`.

**Comment by villebro** [comment]
Sorry, I missed that part. If the full query can't be found in the logs anywhere (looks like it truncates it) and if you can temporarily add a `print(sql)` after the following line in `superset/connectors/sqla/models.py`, then you should be able to get the full query out:

https://github.com/apache/incubator-superset/blob/0e93a94ae4febaef1401aa58ea4481342397a723/superset/connectors/sqla/models.py#L776

**Comment by victornoel** [comment]
@villebro here it is:
```sql
SELECT DATE_TRUNC('day', invoiceDate) AT TIME ZONE 'UTC' AS __timestamp, COUNT(*) AS count FROM public."InvoiceLine" WHERE "invoiceDate" >= '2018-09-10 00:00:00' AND "invoiceDate" <= '2018-09-17 00:00:00' GROUP BY DATE_TRUNC('day', invoiceDate) AT TIME ZONE 'UTC' ORDER BY count DESC LIMIT 10000
```

For the record, I applied this diff to "fix" the problem with a hack but I don't think it is generic at all:
```diff
diff --git a/superset/db_engine_specs.py b/superset/db_engine_specs.py
index 176fbed0..a2ea8cac 100644
--- a/superset/db_engine_specs.py
+++ b/superset/db_engine_specs.py
@@ -397,15 +397,15 @@ class PostgresBaseEngineSpec(BaseEngineSpec):
     engine = ''
 
     time_grain_functions = {
-        None: '{col}',
-        'PT1S': "DATE_TRUNC('second', {col}) AT TIME ZONE 'UTC'",
-        'PT1M': "DATE_TRUNC('minute', {col}) AT TIME ZONE 'UTC'",
-        'PT1H': "DATE_TRUNC('hour', {col}) AT TIME ZONE 'UTC'",
-        'P1D': "DATE_TRUNC('day', {col}) AT TIME ZONE 'UTC'",
-        'P1W': "DATE_TRUNC('week', {col}) AT TIME ZONE 'UTC'",
-        'P1M': "DATE_TRUNC('month', {col}) AT TIME ZONE 'UTC'",
-        'P0.25Y': "DATE_TRUNC('quarter', {col}) AT TIME ZONE 'UTC'",
-        'P1Y': "DATE_TRUNC('year', {col}) AT TIME ZONE 'UTC'",
+        None: '"{col}"',
+        'PT1S': "DATE_TRUNC('second', \"{col}\") AT TIME ZONE 'UTC'",
+        'PT1M': "DATE_TRUNC('minute', \"{col}\") AT TIME ZONE 'UTC'",
+        'PT1H': "DATE_TRUNC('hour', \"{col}\") AT TIME ZONE 'UTC'",
+        'P1D': "DATE_TRUNC('day', \"{col}\") AT TIME ZONE 'UTC'",
+        'P1W': "DATE_TRUNC('week', \"{col}\") AT TIME ZONE 'UTC'",
+        'P1M': "DATE_TRUNC('month', \"{col}\") AT TIME ZONE 'UTC'",
+        'P0.25Y': "DATE_TRUNC('quarter', \"{col}\") AT TIME ZONE 'UTC'",
+        'P1Y': "DATE_TRUNC('year', \"{col}\") AT TIME ZONE 'UTC'",
     }
 
     @classmethod
```

**Comment by villebro** [comment]
@victornoel Fixing this in a native SQL Alchemy way would require writing something along the lines of `sa.sql.func.DATE_TRUNC(text("'second'"), col)` just to get the expression `DATE_TRUNC('second', col)` when compiling the query. This doesn't look too readable IMO. Not to mention having to change how all other time grains are created in `db_engine_specs` for other engines. Unless this turns out to be a more common problem in other engines, I think your fix is probably the best solution, as forcing quotes around the column names should not have any adverse side effects.

**Comment by victornoel** [comment]
@villebro I think there is a misunderstanding. The problem is with `col`, not with the period specifier.

The fix should be quite simple: always put quotes around the column name when generating those query parts.

This is, I believe, the standard SQL way of writing fields, see for example: https://www.jooq.org/doc/3.11/manual-single-page/#settings-name-style.

Now the question is: are all db going to work with that? Maybe you can only fix this for postgresql (which use case sensitivity by default for column names, contrary to, say, sql server) for now, but normally this should be standard SQL.

**Comment by villebro** [comment]
TL;DR: I was talking more about how the query is generated in Superet, not what is valid SQL syntax. I believe that in this instance putting quotes around `{col}` for Postgres is probably the simplest solution, but I would not apply the same to all other engine types.

Long answer: I was referring to how Superset generates SQL queries. Currently most other parts of the query are generated using SQL Alchemy constructs that automatically detect the need for quotes. For example, selecting a column `ColumnName` with an alias `col_alias` is in fact roughly defined as 
`column('ColumnName').label('col_alias')`, which is then compiled into `SELECT "ColumnName" AS col_alias`. If either require quoting, `ColumnName` in this example, SQL Alchemy automatically puts quotes around it.

The problem here is that time grain expressions don't contain a reference to a SQL Alchemy column object, but are static strings inside a `literal_column` object, which makes it impossible for SQL Alchemy to apply conditional quoting. See below where this happens:

https://github.com/apache/incubator-superset/blob/master/superset/connectors/sqla/models.py#L139-L143

I believe that Postgres is pickier than other engines in this case, i.e. most other engines probably wouldn't require quotes around the column name in the time grain expression. Therefore I would not force quotes into the time grain expressions of other engines at this time.

**Comment by victornoel** [comment]
@villebro thanks for the explanation :)
So I will let you decide what is the best answer to this bug then!

**Comment by villebro** [comment]
No problem; if you've made sure it works and feel this is the right course of action, feel free to put through a PR so that the core devs can decide what to do.

**Comment by victornoel** [comment]
@villebro ok, will do :)

**Comment by victornoel** [comment]
FYI, the PR was closed, because, I quote @rumbin:
>I am concerned that this PR will again revive #3844. From the issues and PRs within this very issue you can see that this topic is very controversial.
>
>That time we came to these conclusions:
>
> - Quoting the time column name breaks calculated time columns.
> - A workaround for Case-sensitive time columns may be to rename them by creating a calculated column out of them.

I would be great to tackle this situation properly by stopping generating SQL queries by hand and use the proper tools (i.e. SQLAlchemy) even if it's "less readable".

WDYT @villebro @rumbin @mistercrunch?

**Comment by victornoel** [comment]
For the record, a second workaround was provided here: https://github.com/apache/incubator-superset/pull/5962#issuecomment-425623999.

If it is possible, I would like to go farther than workarounds so that this can be fixed seriously in superset (for the sake of correctness and for future users ;). I can contribute if needed with proper guidance :)
